File: presentation/src/main/java/com/somename/producthunt/screen/presenter/ProductPresenter.java
package com.somename.producthunt.screen.presenter;


import android.support.annotation.NonNull;

import com.somename.domain.model.RootProduct;
import com.somename.domain.usecase.GetProduct;
import com.somename.producthunt.сontent.RootProductViewModel;
import com.somename.producthunt.сontent.mapper.RootProductViewModelMapper;

import javax.inject.Inject;

import io.reactivex.observers.DisposableObserver;

public class ProductPresenter extends Presenter<ProductPresenter.View>{

    private final GetProduct mGetProduct;

    private RootProductViewModelMapper mRootProductViewModelMapper;

    @Inject
    public ProductPresenter(@NonNull GetProduct getProduct, @NonNull RootProductViewModelMapper rootProductViewModelMapper) {
        mGetProduct = getProduct;
        mRootProductViewModelMapper = rootProductViewModelMapper;
    }

    public void init(String id) {
        getView().showLoading();
        mGetProduct.setProductId(id);
        mGetProduct.execute(new DisposableObserver<RootProduct>() {
            @Override
            public void onNext(RootProduct rootProduct) {
                getView().showProduct(mRootProductViewModelMapper.map(rootProduct));
            }

            @Override
            public void onError(Throwable e) {
                getView().hideLoading();
                getView().showError();
                e.printStackTrace();
            }

            @Override
            public void onComplete() {
                getView().hideLoading();
            }
        });
    }

    public void onDestroy() {
        mGetProduct.dispose();
    }

    public interface View extends Presenter.View {

        void showProduct(RootProductViewModel rootProductViewModel);

    }
}

File: presentation/src/main/java/com/somename/producthunt/screen/presenter/ProductPresenter.kt
package com.somename.producthunt.screen.presenter


import com.somename.domain.model.RootProduct
import com.somename.domain.usecase.GetProduct
import com.somename.producthunt.сontent.RootProductViewModel
import com.somename.producthunt.сontent.mapper.RootProductViewModelMapper

import javax.inject.Inject

import io.reactivex.observers.DisposableObserver

class ProductPresenter @Inject
constructor(private val mGetProduct: GetProduct, private val mRootProductViewModelMapper: RootProductViewModelMapper) : Presenter<ProductPresenter.View>() {

    fun init(id: String) {
        view?.showLoading()
        mGetProduct.setProductId(id)
        mGetProduct.execute(object : DisposableObserver<RootProduct>() {
            override fun onNext(rootProduct: RootProduct) {
                view?.showProduct(mRootProductViewModelMapper.map(rootProduct))
            }

            override fun onError(e: Throwable) {
                view?.hideLoading()
                view?.showError()
                e.printStackTrace()
            }

            override fun onComplete() {
                view?.hideLoading()
            }
        })
    }

    fun onDestroy() {
        mGetProduct.dispose()
    }

    interface View : Presenter.View {

        fun showProduct(rootProductViewModel: RootProductViewModel)

    }
}
